sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java

```diff
@@ -2083,4 +2097,174 @@ public String generateSas(BlobServiceSasSignatureValues blobServiceSasSignatureV
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.blob.specialized.BlobAsyncClientBase.setImmutabilityPolicy#BlobImmutabilityPolicy}
+     *
+     * @param immutabilityPolicy {@link BlobImmutabilityPolicy The immutability policy}.
+     * @return A reactive response containing the immutability policy.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<BlobImmutabilityPolicy> setImmutabilityPolicy(BlobImmutabilityPolicy immutabilityPolicy) {
+        try {
+            return setImmutabilityPolicyWithResponse(immutabilityPolicy, null).flatMap(FluxUtil::toMono);
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    /**
+     * Sets the immutability policy on a blob, blob snapshot or blob version.
+     * <p> NOTE: Blob Versioning must be enabled on your storage account and the blob must be in a container with
+     * immutable storage with versioning enabled to call this API.</p>
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.blob.specialized.BlobAsyncClientBase.setImmutabilityPolicyWithResponse#BlobImmutabilityPolicy-BlobRequestConditions}
+     *
+     * @param immutabilityPolicy {@link BlobImmutabilityPolicy The immutability policy}.
+     * @param requestConditions {@link BlobRequestConditions}
+     * @return A reactive response containing the immutability policy.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Response<BlobImmutabilityPolicy>> setImmutabilityPolicyWithResponse(
+        BlobImmutabilityPolicy immutabilityPolicy, BlobRequestConditions requestConditions) {
+        try {
+            return withContext(context -> setImmutabilityPolicyWithResponse(immutabilityPolicy, requestConditions,
+                context));
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    Mono<Response<BlobImmutabilityPolicy>> setImmutabilityPolicyWithResponse(
+        BlobImmutabilityPolicy immutabilityPolicy, BlobRequestConditions requestConditions, Context context) {
+        context = context == null ? Context.NONE : context;
+        BlobImmutabilityPolicy finalImmutabilityPolicy = immutabilityPolicy == null ? new BlobImmutabilityPolicy()
+            : immutabilityPolicy;
+        if (BlobImmutabilityPolicyMode.MUTABLE.equals(finalImmutabilityPolicy.getPolicyMode())) {
+            throw logger.logExceptionAsError(new IllegalArgumentException(
+                String.format("immutabilityPolicy.policyMode must be %s or %s",
+                    BlobImmutabilityPolicyMode.LOCKED.toString(), BlobImmutabilityPolicyMode.UNLOCKED.toString())));
+        }
+
+        BlobRequestConditions finalRequestConditions = requestConditions == null
+            ? new BlobRequestConditions() : requestConditions;
+        // TODO (gapra) : Discuss, should we just expose ifUnmodifiedSince here?
```

Comment:
Let's do this https://github.com/seanmcc-msft/azure-sdk-for-net/pull/2/files for this API since it's brand new.
And let's follow up with adding validation in other places like this (see offline email discussion about this).